src/vsphere_cpi/spec/unit/cloud/vsphere/nsxt_provider_spec.rb

```diff
@@ -194,13 +195,45 @@
                                      .and_return([logical_port_1, logical_port_2])
+          expect(nsxt_provider).to receive(:retrieve_nsgroups).with(ns_groups).and_return([nsgroup_1, nsgroup_2])
         end
 
         it 'adds simple expressions containing the logical ports to each NSGroup' do
-          expect(nsxt_provider).to receive(:retrieve_nsgroups).with(ns_groups).and_return([nsgroup_1, nsgroup_2])
-          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).twice
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_return(true).twice
+          nsxt_provider.add_vm_to_nsgroups(vm, ns_groups)
+        end
+        it 'should raise an error if there is error adding vm to the nsgroup' do
+          failure_response = NSXT::ApiCallError.new(:code => 400)
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_raise(failure_response)
+          expect do
+            nsxt_provider.add_vm_to_nsgroups(vm, ns_groups)
+          end.to raise_error(NSXT::ApiCallError)
+        end
+        it "should retry when there is a CONFLICT in adding vm to the nsgroup" do
+          conflict_response = NSXT::ApiCallError.new(:code => 409, :response_body => 'The object was modified by somebody else')
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_return(true)
+          2.times{ expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_raise(conflict_response) }
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_return(true)
           nsxt_provider.add_vm_to_nsgroups(vm, ns_groups)
         end
+        it "should retry when there is a PreconditionFailed in adding vm to the nsgroup" do
+          precondition_failed_response = NSXT::ApiCallError.new(:code => 412, :response_body => 'PreconditionFailed')
+          4.times{ expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_raise(precondition_failed_response) } #retry 4 times
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_return(true).twice
+          nsxt_provider.add_vm_to_nsgroups(vm, ns_groups)
+        end
+        it "should retry 50 times when there is a conflict or precondition failed while adding vm to nsgroup" do
+          conflict_response = NSXT::ApiCallError.new(:code => 409, :response_body => 'The object was modified by somebody else')
+          retryable = Bosh::Retryable.new()
+          allow(Bosh::Retryable)
+              .to receive(:new)
+                      .with(hash_including(tries: 50))
+                      .and_return(retryable)
+          expect(grouping_obj_svc).to receive(:add_or_remove_ns_group_expression).with(any_args).and_raise(conflict_response)
+          expect do
```

Comment:
Separate the actual test from the mocks with a blank line. There's a lot of setup and it's hard to understand the actual expectation.